### PR TITLE
[RaccOOD] raccoons_OOD.py refactor

### PR DESCRIPTION
**Pull request by kotekjedi**
* imagenet.py created
* model's imagenet evaluation added
* caching added to models/base.py
* tasks/utils.py created
* raccoons_OOD.py fixed

# Task Description
Add a short and general description of your task here. Ideally, also show some samples of the dataset.

## Dataset Creation
Describe the collection/creation of the images present in this task.

## Evaluation Metrics
Which evaluation metrics does your task use? Why are they relevant?

## Expected Insights/Relevance
Which insights do you expect from testing models on this new task/dataset?

# Access

## Data
Add links to access the data here. 

## License
Do you have the permission to release the images in your task? Is the license of the data and your code compatible with the license of this repository?

### CODE REVIEW

File: examples/imagenet_c.py [hunk 1]
```diff
@@ -398,6 +398,8 @@ def _evaluate(self, model: sh_models.Model) -> TaskResult:
 
 
 if __name__ == "__main__":
-    from shifthappens.models.torchvision import resnet18
+    from shifthappens.models.torchvision import ResNet18
 
-    sh_benchmark.evaluate_model(resnet18(), "test_data")
+    sh_benchmark.evaluate_model(
+        ResNet18(device="cuda", max_batch_size=128), "test_data"
```
Comment (zimmerrol) [inline examples/imagenet_c.py:404]: replace `cuda` where with `cpu` such that the code also runs when no CUDA-enabled GPU is available.

File: examples/imagenet_r.py [hunk 2]
```diff
@@ -76,6 +76,6 @@ def _evaluate(self, model: sh_models.Model) -> TaskResult:
 
 
 if __name__ == "__main__":
-    from shifthappens.models.torchvision import resnet18
+    from shifthappens.models.torchvision import ResNet18
 
-    sh_benchmark.evaluate_model(resnet18(), "data")
+    sh_benchmark.evaluate_model(ResNet18(device="cuda", max_batch_size=128), "data")
```
Comment (zimmerrol) [inline examples/imagenet_r.py:81]: replace `cuda` where with `cpu` such that the code also runs when no CUDA-enabled GPU is available.

File: examples/raccoons_OOD.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import dataclasses
```
Comment (zimmerrol) [inline examples/raccoons_OOD.py:1]: Add missing information about the authors of the original dataset, description of the test, etc.

File: examples/raccoons_OOD.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                Metric.Robustness: "accuracy",  # remove for pure OOD detection
+            },
+        )
+
+
+if __name__ == "__main__":
+    from shifthappens.models.torchvision import ResNet18
+
+    sh_benchmark.evaluate_model(
+        ResNet18(device="cuda", max_batch_size=128),
```
Comment (zimmerrol) [inline examples/raccoons_OOD.py:100]: replace `cuda` where with `cpu` such that the code also runs when no CUDA-enabled GPU is available.

File: shifthappens/data/imagenet.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
+"""
+This module provides functionality to access ImageNet validation set ground truth targets,
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:2]: Update comment:

This module provides functionality to access the predictions of a model for the ImageNet validation set. Futher, the predictions can be cached and loaded from the cache to reduce computational costs. Note, that for this to work
:py:data:`shifthappens.data.imagenet.ImageNetValidationData` must be set to the ImageNet validation set path.

File: shifthappens/data/imagenet.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import shutil
+
+import numpy as np
+import torchvision.datasets as tv_datasets
+import torchvision.transforms as tv_transforms
+
+import shifthappens.data.torch as sh_data_torch
+from shifthappens.data.base import DataLoader
+
+#: Must be set to ImageNet validation set PATH.
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:19]: Change to `#: Must be set to ImageNet validation set path (either absolute or relative to working directory).`

File: shifthappens/data/imagenet.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+ImageNetValidationData = "shifthappens/imagenet"
+
+#: Must be set to models' results caching directory.
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:22]: Change to `#: Must be set to models' results caching directory (either absolute or relative to working directory). If the folder does not exist, it will be created.`

File: shifthappens/data/imagenet.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+ImageNetValidationPredictionsCache = "shifthappens/cache"
+
+
+def _check_imagenet_folder():
+    """
+    Checks if path to ImageNet folder is defined, the folder exists and contains
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:28]: Change `Checks if path to the ImageNet validation set folder is defined, the folder exists and contains`

File: shifthappens/data/imagenet.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    a thousand folders.
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:29]: Change to `a thousand folders (one per class).`

File: shifthappens/data/imagenet.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    """
+    assert (
+        ImageNetValidationData is not None
+    ), "ImagenetValidationData path is not specified."
+    assert os.path.exists(
+        ImageNetValidationData
+    ), "You have specified incorrect path to ImageNet. Files not found at the ImageNetValidationData path."
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:36]: Chagne to `"You have specified an incorrect path to the ImageNet validation set. Files not found at location specified in shithappens.data.imagenet.ImageNetValidationData."`

File: shifthappens/data/imagenet.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+    assert (
+        len(os.listdir(ImageNetValidationData)) == 1000
+    ), "ImageNetValidationData folder contains less or more folders then ImageNet classes."
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:40]: `then` should be `than`

File: shifthappens/data/imagenet.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+
+def get_imagenet_val_dataloader(max_batch_size=128) -> DataLoader:
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:43]: Rename function to `get_imagenet_validation_loader`

File: shifthappens/data/imagenet.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    """
+    Returns ImageNet validation set DataLoader. Note
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:45]: Change to `Creates a DataLoader for the validation set of ImageNet`

Also, DataLoader in this comment should be replaced with the correct :py:class: link.

File: shifthappens/data/imagenet.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    Args:
+        cls: Model's class. Used for specifying folder name.
+
+    Returns:
+        Dictionary of loaded model predictions on ImageNet validation set.
+    """
+    assert (
+        ImageNetValidationPredictionsCache is not None
+    ), "Cannot get cached model results. ImageNetValidationPredictionsCache path is not specified."
+    load_path = ImageNetValidationPredictionsCache + "/" + cls.__class__.__name__ + "/"
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:90]: use `os.path.join`

File: shifthappens/data/imagenet.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    imagenet_val_dataloader = DataLoader(
+        imagenet_val_dataset, max_batch_size=max_batch_size
+    )
+
+    return imagenet_val_dataloader
+
+
+def get_cached_predictions(cls) -> dict:
+    """
+    Looks up in cache directory for a cls-named folder and load
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:77]: Change this to `Checks whether there exist cached results for the model's class and if so, returns them`

File: shifthappens/data/imagenet.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    )
+
+    imagenet_val_dataloader = DataLoader(
+        imagenet_val_dataset, max_batch_size=max_batch_size
+    )
+
+    return imagenet_val_dataloader
+
+
+def get_cached_predictions(cls) -> dict:
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:75]: It should be defined what happens if the function is called for a class for which no results have been cached before, i.e. an exception should be raised.

File: shifthappens/data/imagenet.py [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    for file in os.listdir(load_path):
+        result = np.load(load_path + file)
+        result_dict[file.rstrip(".npy")] = result
+    return result_dict
+
+
+def cache_predictions(cls, imagenet_validation_result):
+    """
+    Caches model predictions in cls-named folder and load
+    model predictions from it. Note that path to ImageNet validation set
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:101]: That doesn't make sense to me: why do we first store the results and then _load_ them again?

File: shifthappens/data/imagenet.py [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+    Args:
+        cls: Model's class. Used for specifying folder name.
+        imagenet_validation_result (ModelResult): Model's prediction on ImageNet
+            validation set.
+    """
+
+    assert (
+        ImageNetValidationPredictionsCache is not None
+    ), "Cannot cache model results. ImageNetValidationPredictionsCache path is not specified."
+    save_path = ImageNetValidationPredictionsCache + "/" + cls.__class__.__name__ + "/"
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:114]: see above; use `os.path.join`

File: shifthappens/data/imagenet.py [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+
+    if os.path.exists(save_path):
+        shutil.rmtree(save_path)
+    os.makedirs(save_path)
+
+    for result_type in imagenet_validation_result.__slots__:
+        result = getattr(imagenet_validation_result, str(result_type))
+        if result is not None:
+            np.save(save_path + str(result_type), result)
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:123]: use `os.path.join`

File: shifthappens/data/imagenet.py [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    Returns:
+        Dictionary of loaded model predictions on ImageNet validation set.
+    """
+    assert (
+        ImageNetValidationPredictionsCache is not None
+    ), "Cannot get cached model results. ImageNetValidationPredictionsCache path is not specified."
+    load_path = ImageNetValidationPredictionsCache + "/" + cls.__class__.__name__ + "/"
+    result_dict = dict()
+    for file in os.listdir(load_path):
+        result = np.load(load_path + file)
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:93]: use `os.path.join`

File: shifthappens/data/imagenet.py [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+    Args:
+        cls: Model's class. Used for specifying folder name.
+
+    Returns:
+        ``True`` if model's results are cached, ``False`` otherwise.
+    """
+    assert (
+        ImageNetValidationPredictionsCache is not None
+    ), "Cannot find cached model results. ImageNetValidationPredictionsCache path is not specified."
+    load_path = ImageNetValidationPredictionsCache + "/" + cls.__class__.__name__ + "/"
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:142]: use `os.path.join`

File: shifthappens/data/imagenet.py [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+    for result_type in imagenet_validation_result.__slots__:
+        result = getattr(imagenet_validation_result, str(result_type))
+        if result is not None:
+            np.save(save_path + str(result_type), result)
+
+
+def is_cached(cls) -> bool:
+    """
+    Checks if model's results are cached in cls-named folder. Note
+    that path to ImageNet validation set
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:129]: `Note that the path to the ImageNet validation set`

File: shifthappens/data/imagenet.py [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+        cached_files = os.listdir(load_path)
+    except FileNotFoundError:
+        print(f"There is no cached model results on ImageNet at {load_path}.")
+        return False
+    return True
+
+
+def load_imagenet_targets() -> np.ndarray:
+    """
+    Returns ground truth targets of ImageNet validations set.
```
Comment (zimmerrol) [inline shifthappens/data/imagenet.py:154]: Change to `Returns the ground-truth labels of the ImageNet valdation set.`

File: shifthappens/models/base.py [hunk 24]
```diff
@@ -106,6 +107,22 @@ class Model(abc.ABC):
 
     """
 
+    _imagenet_validation_result = ModelResult(np.array([]), None, None, None)
+
+    @property
+    def imagenet_validation_result(self) -> ModelResult:
+        """
+        Getter-like property to access model's evaluation result on ImageNet validation set.
```
Comment (zimmerrol) [inline shifthappens/models/base.py:115]: Change to `Access the model's predictions/evaluation results on the ImageNet validation set.`

File: shifthappens/models/base.py [hunk 25]
```diff
@@ -106,6 +107,22 @@ class Model(abc.ABC):
 
     """
 
+    _imagenet_validation_result = ModelResult(np.array([]), None, None, None)
```
Comment (zimmerrol) [inline shifthappens/models/base.py:110]: Move this to the `__init__` method; also, initialize this as `None` and not a `ModelResult` with no results in it.

File: shifthappens/models/base.py [hunk 26]
```diff
@@ -106,6 +107,22 @@ class Model(abc.ABC):
...
+
+    @property
+    def imagenet_validation_result(self) -> ModelResult:
+        """
+        Getter-like property to access model's evaluation result on ImageNet validation set.
+
+        Returns:
+            Model evaluation result on ImageNet validation set wrapped with ModelResult.
+
```
Comment (zimmerrol) [inline shifthappens/models/base.py:119]: Remove blank line.

File: shifthappens/models/base.py [hunk 27]
```diff
@@ -106,6 +107,22 @@ class Model(abc.ABC):
...
+        """
+        if len(self._imagenet_validation_result.class_labels) == 0:
```
Comment (zimmerrol) [inline shifthappens/models/base.py:121]: Change to `if self._imagenet_validation_result.class_labels is None:`

**Review by zimmerrol — CHANGES_REQUESTED** [review]

File: shifthappens/data/imagenet.py (hunk 17 above)
Comment (kotekjedi) [reply shifthappens/data/imagenet.py]: This is how I see it:

when we refer to the model's prediction on ImageNet for the first time, and if there are no cached results yet, we do the following: model predicts in ImageNet, and we cache the results. (No loading called here)

If we reinitialize this model or rerun the whole benchmark - we load the results from the cache folder.

So we don't use store and load functions sequentially in the one run for a model

**Review by zimmerrol — APPROVED** [review]
LGTM
